diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -662,6 +676,27 @@ const Value *DifferentialFunctionComparator::getReplacementValue(
     return replacementIt->second;
 }
 
+/// Creates new value mappings according to the current pattern match.
+void DifferentialFunctionComparator::createPatternMapping() const {
+    for (auto &&MappedInstPair : PatternComp.InstMappings) {
+        snPairMap[sn_mapL.size()] = {MappedInstPair.first,
+                                     MappedInstPair.second};
+        sn_mapL[MappedInstPair.first] = sn_mapL.size();
```

Comment:
`llvm:DenseMap` only uses `insert` (which internally uses `try_emplace`).

Regarding the mapping, it is semi-guaranteed (i.e., no, not really). The algorithm looks ahead in the compared functions, so it should always create new mappings. However, if multiple patterns match in a single function, duplicates (and related issues) might arise.

In general, this scenario is not supported by the current implementation, so I suppose that a check here that ensures that only the first mapping gets used in case of duplicate mappings would be a great idea (almost certainly a bugfix, actually).

Reply:
Yeah, `try_emplace` is good as well. In any case, it wouldn't overwrite the value and would allow you to check if a new mapping was inserted.